src/input_common.cpp

```diff
@@ -270,3 +270,125 @@ void input_event_queue_t::prepare_to_select() {}
 void input_event_queue_t::select_interrupted() {}
 void input_event_queue_t::uvar_change_notified() {}
 input_event_queue_t::~input_event_queue_t() = default;
+
+readline_cmd_t::readline_cmd_t(const readline_cmd_t& rl) : id_(rl.id_) {
+    if (id_ == id_t::insert_chars)
+        new (&arg_.insertion) commandline_insertion_t(rl.arg_.insertion);
+    else if (id_ == id_t::set_cursor)
+        arg_.cursor_pos = rl.arg_.cursor_pos;
+}
+
+readline_cmd_t::readline_cmd_t(readline_cmd_t&& rl) : id_(rl.id_) {
+    if (id_ == id_t::insert_chars)
+        new (&arg_.insertion) commandline_insertion_t(std::move(rl.arg_.insertion));
+    else if (id_ == id_t::set_cursor)
+        arg_.cursor_pos = rl.arg_.cursor_pos;
+}
+
+readline_cmd_t& readline_cmd_t::operator=(const readline_cmd_t& rl) {
+    if (this != &rl) {
+        // commandline_insertion_t is a non-aggregate type so needs to
+        // initialized it when it becomes the active union member
+        // freed it when it is deactivated.
+        if (id_ == id_t::insert_chars) {
+            if (rl.id_ == id_t::insert_chars)
+                arg_.insertion = rl.arg_.insertion;
+            else
+                arg_.insertion.~commandline_insertion_t();
+        } else if (rl.id_ == id_t::insert_chars) {
+            new (&arg_.insertion) commandline_insertion_t(rl.arg_.insertion);
+        }
+        if (rl.id_ == id_t::set_cursor) arg_.cursor_pos = rl.arg_.cursor_pos;
+        id_ = rl.id_;
+    }
+    return *this;
+}
+
+readline_cmd_t& readline_cmd_t::operator=(readline_cmd_t&& rl) {
+    if (this != &rl) {
+        // commandline_insertion_t is a non-aggregate type so needs to
+        // initialized it when it becomes the active union member
+        // freed it when it is deactivated.
+        if (id_ == id_t::insert_chars) {
+            if (rl.id_ == id_t::insert_chars)
+                arg_.insertion = std::move(rl.arg_.insertion);
+            else
+                arg_.insertion.~commandline_insertion_t();
+        } else if (rl.id_ == id_t::insert_chars) {
+            new (&arg_.insertion) commandline_insertion_t(std::move(rl.arg_.insertion));
+        }
+        if (rl.id_ == id_t::set_cursor) arg_.cursor_pos = rl.arg_.cursor_pos;
+        id_ = rl.id_;
+    }
+    return *this;
+}
+
+readline_cmd_t::~readline_cmd_t() {
+    // Only destroy arg_.insertion if it is active.
+    if (id_ == id_t::insert_chars) arg_.insertion.~commandline_insertion_t();
+}
```

Comment:
this is tricky code but it looks correct.

> The fourth commit should perhaps be squashed into the fifth to avoid having to comment out the temporary test failures

I think so too, although it was nice to read this way

Reply:
I'll squash them together.